Removing special chars that are in the end of a url and shouldn't be

ShogunFire opened the pull request:
Because we are removing unmatched end parenthesis, some links that end with special characters are accepted when they shouldn't be. we need to remove those special characters.

### Fixed Issues
$ https://github.com/Expensify/App/issues/25532

# Tests
1. Go to any repórt
2. Send a link with a slash for example: google.com/toto
3. Edit the message and add one of those characters to the link in the parenthese: `$*.+!,=`
4. When you validate the edit verify that the text is white without links like this:
![image](https://github.com/Expensify/expensify-common/assets/19537677/2fbfb1e5-bba6-47f2-ba4d-cd10ecb65c50)

5. Send another message google.com/toto.) or google.com/toto$) for example and verify that the special char before the parenthese is not blue (not included in the link)
![image](https://github.com/Expensify/expensify-common/assets/19537677/7f1dad72-4cd7-4cdf-830a-6af41717c90e)


# QA
Same as above and test links in general

Video for web:


https://github.com/Expensify/expensify-common/assets/19537677/8bd1d2ca-5695-412a-8690-448a46295b4c

ShogunFire commented:
I have read the CLA Document and I hereby sign the CLA

ShogunFire commented:
@0xmiroslav

ShogunFire commented:
Ok I am adding test cases for autolink url with special chars and unmatched parenthese in ExpensiMark-HTML-test file, I think that should be enough.

ShogunFire commented:
I added the video and some test cases

MonilBhavsar commented:
Requesting a review

ShogunFire commented on lib/ExpensiMark.js:
```diff
@@ -468,6 +468,24 @@ export default class ExpensiMark {
+                match[0] = match[0].substr(0, match[0].length - numberOfCharsToRemove);
+                match[2] = match[2].substr(0, match[2].length - numberOfCharsToRemove);
+            }
+
```
There is probably a way to make that work in the same code but I don't think it makes sense to try to do it because the code for `)` has to count the parentheses in the left to right order while this code starts from the end.
Also this situation where we have the special chars at the end only happens after we have remove the `)`

I am adding the special characters in CONST right now

ShogunFire commented:
@0xmiroslav The special chars are in the const file now

ShogunFire commented:
@0xmiroslav I made the suggestions

MonilBhavsar commented on lib/ExpensiMark.js:
```diff
@@ -468,6 +468,24 @@ export default class ExpensiMark {
                 }
             }
+
+            // Because we are removing ) parenthese, some special characters that shouldn't be in the href are in the href
```
```suggestion
            // Because we are removing ) parenthesis, some special characters that shouldn't be in the href are in the href
```

MonilBhavsar commented on lib/ExpensiMark.js:
```diff
@@ -468,6 +468,24 @@ export default class ExpensiMark {
             }
+
+            // Because we are removing ) parenthese, some special characters that shouldn't be in the href are in the href
+            // For example google.com/toto.) is accepted by the regexp, above we remove the ) parenthese so the link becomes google.com/toto. which is not a valid link
```
```suggestion
            // For example google.com/toto.) is accepted by the regular expression above and we remove the ) parenthesis, so the link becomes google.com/toto. which is not a valid link
```

MonilBhavsar reviewed (commented):
Small comment. Looks good. Thanks for adding the tests!

MonilBhavsar commented:
@ShogunFire the Test Javascript is failing for some reason. Could you please merge main and check

ShogunFire commented:
Ok so it looks like I was wrong about those special characters never being at the end of an url. In fact those characters are never at the end of a `URL_PATH_REGEX` but they can be at the end of strings that are after the path like `URL_PARAM_REGEX` or `URL_FRAGMENT_REGEX` here: https://github.com/Expensify/expensify-common/blob/cb1c6b75cb593990ed18381ba7dba8e8ceb9f16b/lib/Url.js#L7-L9

In their tests they have `=` that is at the end and it is valid because it is part of the parameter of the url. One way to resolve this is that before I remove the special chars that are at the end I must test that we don't have a param or a fragment by testing if we have a `?` or `#` in the url, do you think this would be the right way ?

ShogunFire commented:
Hello @0xmiroslav Can you check this: https://github.com/Expensify/expensify-common/pull/575#issuecomment-1727814421

ShogunFire commented:
With the current regexp we have it's not only = that can be at the end, all those characters `$*.+!,` can also be at the end if we have a param, I don't think we should remove them either if they are part of a param.

And never removing = will create an inconsistency: `google.com/toto=)` the = will be part of the url while in `google.com/toto=` it won't be in the url

ShogunFire commented:
Without `?` this is not a param I think

ShogunFire commented:
Yes, this is what I propose:

> One way to resolve this is that before I remove the special chars that are at the end I must test that we don't have a param or a fragment by testing if we have a ? or # in the url

MonilBhavsar commented:
I also agree with the proposal 👍 
Just curious how we're thinking to implement it?

ShogunFire commented:
@0xmiroslav @MonilBhavsar I commited, I just tested like this:
` if (!url.includes('?') && !url.includes('#')) {`
is it enough ? 

I merged with main and replaced the `match[2]` with `url` since we created this variable in the other PR.

I added tests for params and fragments

MonilBhavsar commented:
I think so it will work. What could be the other cases though 🤔 
@0xmiroslav could you please take a look at the comment above?

ShogunFire commented:
@0xmiroslav This question was for me ? I don't know but it was like this in the regexp, I think the regexp might not be perfect but we can see that on different websites (github and slack for example) they have different rules so I would leave it like this just in case we are missing something

ShogunFire commented:
@0xmiroslav Yes I think we are good now

ShogunFire commented:
@0xmiroslav I pulled main

MonilBhavsar reviewed (approved):
Thanks! Sorry, I missed it somehow

MonilBhavsar commented:
I'm pretty sure, the tests were passing

<img width="1728" alt="Screenshot 2023-10-17 at 11 36 54 AM" src="https://github.com/Expensify/expensify-common/assets/32012005/a770a065-19d9-4a37-9c8b-75a7abe4aa9a">
